File: Code/js/Lesson_14_code/src/cachingResolver.ts
// A stub resolver with its own cache, keyed by TTL.
//
// The Python twin is caching_resolver.py — same cache shape, same
// expiry math. Recognition, not new material.
//
//   node cachingResolver.ts bookmarks-api.local

import dgram from "node:dgram";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const SERVER_PORT = 5301;
const CACHE_FILE = path.join(
  path.dirname(fileURLToPath(import.meta.url)),
  "resolverCache.json",
);

function encodeName(name: string): Buffer {
  const parts: Buffer[] = [];
  for (const label of name.split(".")) {
    parts.push(Buffer.from([label.length]), Buffer.from(label, "ascii"));
  }
  parts.push(Buffer.from([0]));
  return Buffer.concat(parts);
}

function buildQuery(name: string): Buffer {
  const header = Buffer.from([0x12, 0x34, 0x01, 0x00, 0, 1, 0, 0, 0, 0, 0, 0]);
  const question = Buffer.concat([encodeName(name), Buffer.from([0, 1, 0, 1])]);
  return Buffer.concat([header, question]);
}

async function queryAuthoritative(name: string): Promise<{ ip: string; ttl: number }> {
  const socket = dgram.createSocket("udp4");
  const response = await new Promise<Buffer>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error("timed out")), 2000);
    socket.once("message", (msg) => {
      clearTimeout(timer);
      resolve(msg);
    });
    socket.send(buildQuery(name), SERVER_PORT, "127.0.0.1");
  });
  socket.close();
  const ancount = response.readUInt16BE(6);
  if (ancount === 0) throw new Error(`NXDOMAIN: ${name}`);
  let offset = 12;
  while (response[offset] !== 0) offset += response[offset] + 1;
  offset += 1 + 4; // end of question name + qtype/qclass
  offset += 2; // compression pointer in the answer's name
  const ttl = response.readUInt32BE(offset + 4);
  const rdlength = response.readUInt16BE(offset + 8);
  offset += 10;
  const ip = Array.from(response.subarray(offset, offset + rdlength)).join(".");
  return { ip, ttl };
}

function loadCache(): Record<string, { ip: string; expiresAt: number }> {
  if (fs.existsSync(CACHE_FILE)) return JSON.parse(fs.readFileSync(CACHE_FILE, "utf8"));
  return {};
}

function saveCache(cache: Record<string, { ip: string; expiresAt: number }>): void {
  fs.writeFileSync(CACHE_FILE, JSON.stringify(cache));
}

async function resolve(name: string): Promise<string> {
  const cache = loadCache();
  const now = Date.now() / 1000;
  const entry = cache[name];
  if (entry && entry.expiresAt > now) {
    const remaining = entry.expiresAt - now;
    console.log(`cache hit: ${name} -> ${entry.ip} (${remaining.toFixed(1)}s left on TTL)`);
    return entry.ip;
  }
  const { ip, ttl } = await queryAuthoritative(name);
  cache[name] = { ip, expiresAt: now + ttl };
  saveCache(cache);
  console.log(`cache miss: queried authoritative, ${name} -> ${ip} (ttl=${ttl}s)`);
  return ip;
}

resolve(process.argv[2]);

File: Code/Lesson_14_code/caching_resolver.py
"""A stub resolver with its own cache, keyed by TTL.

A browser, an OS, and every resolver in between each keep this same kind
of cache. This one is small enough to read start to finish: look up a
name; if a cached answer exists and has not expired, return it and send
no packet at all; otherwise query the authoritative server, store the
answer with an expiry computed from the TTL it returned, and return that.

    python caching_resolver.py bookmarks-api.local
    python caching_resolver.py bookmarks-api.local        # second call, same process: cache hit
"""

from __future__ import annotations

import json
import socket
import struct
import sys
import time
from pathlib import Path

SERVER = ("127.0.0.1", 5300)
CACHE_FILE = Path(__file__).parent / "resolver_cache.json"


def encode_name(name: str) -> bytes:
    out = bytearray()
    for label in name.rstrip(".").split("."):
        out += bytes([len(label)]) + label.encode("ascii")
    out += b"\x00"
    return bytes(out)


def build_query(name: str) -> bytes:
    header = struct.pack(">HHHHHH", 0x1234, 0x0100, 1, 0, 0, 0)
    question = encode_name(name) + struct.pack(">HH", 1, 1)  # type A, class IN
    return header + question


def query_authoritative(name: str) -> tuple[str, int]:
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.settimeout(2)
        sock.sendto(build_query(name), SERVER)
        response, _ = sock.recvfrom(512)
    ancount = struct.unpack(">H", response[6:8])[0]
    if ancount == 0:
        raise LookupError(f"NXDOMAIN: {name}")
    # Skip the header (12) and the echoed question (name + 4 bytes of
    # type/class); the answer's name is a compression pointer (2 bytes).
    offset = 12
    while response[offset] != 0:
        offset += response[offset] + 1
    offset += 1 + 4
    offset += 2  # compression pointer
    _, _, ttl, rdlength = struct.unpack(">HHIH", response[offset : offset + 10])
    offset += 10
    ip = socket.inet_ntoa(response[offset : offset + rdlength])
    return ip, ttl


def load_cache() -> dict:
    if CACHE_FILE.exists():
        return json.loads(CACHE_FILE.read_text())
    return {}


def save_cache(cache: dict) -> None:
    CACHE_FILE.write_text(json.dumps(cache))


def resolve(name: str) -> str:
    cache = load_cache()
    now = time.time()
    entry = cache.get(name)
    if entry and entry["expires_at"] > now:
        remaining = entry["expires_at"] - now
        print(f"cache hit: {name} -> {entry['ip']} ({remaining:.1f}s left on TTL)")
        return entry["ip"]
    ip, ttl = query_authoritative(name)
    cache[name] = {"ip": ip, "expires_at": now + ttl}
    save_cache(cache)
    print(f"cache miss: queried authoritative, {name} -> {ip} (ttl={ttl}s)")
    return ip


if __name__ == "__main__":
    resolve(sys.argv[1])
